### ISSUE TITLE
Netopeer2 crashing when compiled from devel branch

### ISSUE THREAD

**Issue by ChinmayaAgarwal** [sysrepo/sysrepo#2820]
Hi @michalvasko 

As part of investigating #2459 , we found another issue #2818 that states that some fix has been done for handling parallel get RPCs. To pick that change we compiled devel branch for packages netopeer2, sysrepo, libyang and libnetconf2. After that we executed the same test mentioned in #2459 i.e. Client creating 6 netconf connection with  6 ip addresses on the machine running netopeer2 server and sending periodic get and create-subscription RPCs.  But we still see that netopeer2 server is crashing but this time with a different back trace:-

```
[DBG]: LN: Session 25: Received message:
<?xml version="1.0" encoding="UTF-8"?>
<hello xmlns="urn:ietf:params:xml:ns:netconf:base:1.0">
  <capabilities>
    <capability>urn:ietf:params:netconf:base:1.0</capability>
    <capability>urn:ietf:params:netconf:base:1.1</capability>
  </capabilities>
</hello>


[INF]: SR: Session 10690 (user "root", CID 25) created.
[DBG]: LN: Session 26: Received message:
<?xml version="1.0" encoding="UTF-8"?>
<hello xmlns="urn:ietf:params:xml:ns:netconf:base:1.0">
  <capabilities>
    <capability>urn:ietf:params:netconf:base:1.0</capability>
    <capability>urn:ietf:params:netconf:base:1.1</capability>
  </capabilities>
</hello>


[INF]: SR: Session 10691 (user "root", CID 25) created.

Thread 9 "netopeer2-serve" received signal SIGABRT, Aborted.
[Switching to Thread 0x7fffdbfff700 (LWP 2148)]
0x00007ffff6b5f37f in raise () from /lib64/libc.so.6
Missing separate debuginfos, use: yum debuginfo-install keyutils-libs-1.5.10-9.el8.x86_64 krb5-libs-1.18.2-14.el8.x86_64 libblkid-2.32.1-27.el8.x86_64 libcap-2.26-4.el8.x86_64 libcom_err-1.45.6-2.el8.x86_64 libgcc-8.5.0-4.el8_5.x86_64 libmount-2.32.1-27.el8.x86_64 libselinux-2.9-5.el8.x86_64 libuuid-2.32.1-27.el8.x86_64 libxcrypt-4.1.1-6.el8.x86_64 openssl-libs-1.1.1k-5.el8_5.x86_64 pcre2-10.32-2.el8.x86_64 sssd-client-2.4.0-9.el8.x86_64 systemd-libs-239-45.el8.x86_64 zlib-1.2.11-17.el8.x86_64
(gdb) bt
#0  0x00007ffff6b5f37f in raise () from /lib64/libc.so.6
#1  0x00007ffff6b49db5 in abort () from /lib64/libc.so.6
#2  0x00007ffff6b49c89 in __assert_fail_base.cold.0 () from /lib64/libc.so.6
#3  0x00007ffff6b57a76 in __assert_fail () from /lib64/libc.so.6
#4  0x00007ffff6f3c16d in sr_shmsub_multi_listen_write_event (multi_sub_shm=0x7ffff7fba000, valid_subscr_count=4,
    err_code=SR_ERR_OK, shm_data_sub=0x7fffdbffebb0, data=0x0, data_len=0, result_str=0x7ffff6f560d8 "Successful")
    at /home/ansible/sysrepo/src/shm_sub.c:2344
#5  0x00007ffff6f3f2ca in sr_shmsub_notif_listen_process_module_events (notif_subs=0x7fffe004c100, conn=0x62cbf0)
    at /home/ansible/sysrepo/src/shm_sub.c:3581
#6  0x00007ffff6efeaf6 in sr_subscription_process_events (subscription=0x7fffe0003500, session=0x0,
    stop_time_in=0x7fffdbffecd0) at /home/ansible/sysrepo/src/sysrepo.c:4001
#7  0x00007ffff6f3f995 in sr_shmsub_listen_thread (arg=0x7fffe0003500) at /home/ansible/sysrepo/src/shm_sub.c:3771
#8  0x00007ffff74d417a in start_thread () from /lib64/libpthread.so.0
#9  0x00007ffff6c24dc3 in clone () from /lib64/libc.so.6
(gdb)
```
Can you please suggest how to resolve this issue?

Thanks and Regards,
Chinmaya Agarwal.

**Comment by ChinmayaAgarwal** [comment]
Hi,

In our scenario, client is creating 6 netconf connections (with 6 different ip address on management interface) with the same netopeer2 server running on Centos 8.4 VM. After every 30 seconds, client re-establishes netconf connection with the server and sends create-subscription and get RPCs after that. This is happening for all the six netconf connections in parallel. This is where we see netopeer2 server crashing after sometime.

Thanks and Regards,
Chinmaya Agarwal.

**Comment by michalvasko** [comment]
Actually, it seems you are not really using *sysrepo* `devel` because according to the stack trace
```
#3  0x00007ffff6b57a76 in __assert_fail () from /lib64/libc.so.6
#4  0x00007ffff6f3c16d in sr_shmsub_multi_listen_write_event (multi_sub_shm=0x7ffff7fba000, valid_subscr_count=4,
    err_code=SR_ERR_OK, shm_data_sub=0x7fffdbffebb0, data=0x0, data_len=0, result_str=0x7ffff6f560d8 "Successful")
    at /home/ansible/sysrepo/src/shm_sub.c:2344
```
there should be an `assert()` on `shm_sub.c:2344` but [there is not](https://github.com/sysrepo/sysrepo/blob/devel/src/shm_sub.c#L2344).

**Comment by ChinmayaAgarwal** [comment]
Hi @michalvasko 

There is no older version running on the machine because I created fresh VM for devel branch compilation. Also, below is the output for the command "sudo ldconfig -p | grep sysrepo":-

[ansible@localhost ~]$ sudo ldconfig -p | grep sysrepo
        libsysrepo.so.7 (libc6,x86-64) => /usr/local/lib64/libsysrepo.so.7
        libsysrepo.so (libc6,x86-64) => /usr/local/lib64/libsysrepo.so
[ansible@localhost ~]$

What is the expectation here? Are we missing anything?

Thanks and Regards,
Chinmaya Agarwal.

**Comment by michalvasko** [comment]
You need to look at `/usr/local/lib64/libsysrepo.so.7` because it is just a symlink, the target should be `/usr/local/lib64/libsysrepo.so.7.8.0` because that is the current `devel`. You know you are using it if the stack trace source code lines match the actual `devel` code.
